[DI] Documenting Abstract Bundle and Extension (#16801)
symfony/symfony-docs: bundles/extension.rst

    {
        $loader = new XmlFileLoader(
            $container,

>             new FileLocator(__DIR__.'/../config')

Author: Do you pretend to upgrade the directory structure? In that case, the sentences above should be updated too.

Reviewer: Oh, you're right. Yes, let's use the new bundle structure everywhere in the 6.1 docs :)

Author: Updated, also the path here is missing one `/../`as the extension class in this case is placed in `src/DependencyInjection/`

    This technique can't be used when the classes to compile use the ``__DIR__``
    or ``__FILE__`` constants, because their values will change when loading
    these classes from the ``classes.php`` file.

Loading Services directly in your Bundle class
----------------------------------------------

.. versionadded:: 6.1

    The ``AbstractBundle`` class is introduced in Symfony 6.1.

Alternatively, you can define and load services configuration directly in a bundle class
by extending from the :class:`Symfony\\Component\\HttpKernel\\Bundle\\AbstractBundle`
and defining the :method:`Symfony\\Component\\HttpKernel\\Bundle\\AbstractBundle::loadExtension` method::

    use Symfony\Component\HttpKernel\Bundle\AbstractBundle;
    use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

    class AcmeFooBundle extends AbstractBundle
    {

>         public function loadExtension(array $config, ContainerConfigurator $container, ContainerBuilder $builder): void

Reviewer 2: This comment is about the feature itself (maybe too late now), but how getting the builder here can be useful? What can be done with it that cannot be done with the configurator?

Author: The `ContainerBuilder` contains still useful methods that are not included in the `ContainerConfigurator`, for example, `removeDefinition()`, `hasDefinition()`, `registerForAutoconfiguration()`, `prependExtensionConfig()`, `getExtensionConfig()`, etc. (the last two useful in the new `prependExtension()`) 

In this method, we used to import the service definitions from an external file, and then we can remove some if desired using the builder instance. That's why we kept both.

Reviewer 2: I see, thank you for your explanation :).

        {
            $container->parameters()
                ->set('foo', $config['foo']);

            $container->import('../config/services.php');

            if ('bar' === $config['foo']) {
                $container->services()
                    ->set(Parser::class);
            }
        }
    }

This method is a shortcut of the previous "load()" method, but with more options
to define and import the service configuration with less effort. The ``$config``

> argument is the previous ``$configs`` array but already merged and processed. And

Reviewer 2: Did we somehow forget to mention https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/HttpKernel/DependencyInjection/ConfigurableExtension.php#L41?

Author: The `ConfigurableExtension` class is not the one used in this feature, in fact, this new feature replaces that class, and I guess we can deprecate it in future versions. https://github.com/symfony/symfony/pull/43701#discussion_r825310122

through the ``$container`` configurator you can import the services configuration
from an external file in any supported format (php, yaml, xml) or simply define
them in place using the fluent interfaces.

.. note::

>     The "loadExtension()" like "load()" method is called only at compiler time.

Reviewer 2: `compile time` as well?

>     The "loadExtension()" like "load()" method is called only at compiler time.

Reviewer 2: ```suggestion
    The "loadExtension()", as the "load()" method, are called only at compile time.
```
